### PR TITLE
Issue 5530: Use grpc-netty-shaded to avoid conflict with other Netty versions

### LINKED ISSUE #5530: Use grpc-netty-shaded to avoid conflict with other Netty version
**Problem description**
Use grpc-netty-shaded to avoid conflict with other Netty versions in the application stack.

**Problem location**
controller-api-client

**Suggestions for an improvement**
use grpc-netty-shaded.

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**  
Use grpc-netty-shaded to avoid conflict with other Netty version.

**Purpose of the change**  
Fixes #5530 

**What the code does**  
Use grpc-netty-shaded to avoid conflict with other Netty version. 

* The dependencies for Pravega-client will not pull netty transport.
Dependency tree for `shared:controller-api`
```
// New dependency tree
+--- project :shared:controller-api
|    +--- org.slf4j:slf4j-api:1.7.25
|    +--- io.grpc:grpc-netty-shaded:1.24.2
|    |    \--- io.grpc:grpc-core:1.24.2
|    |         +--- io.grpc:grpc-api:1.24.2
|    |         |    +--- io.grpc:grpc-context:1.24.2
|    |         |    +--- com.google.errorprone:error_prone_annotations:2.3.3 -> 2.3.4
|    |         |    +--- com.google.code.findbugs:jsr305:3.0.2
|    |         |    +--- org.codehaus.mojo:animal-sniffer-annotations:1.17
|    |         |    \--- com.google.guava:guava:26.0-android -> 28.2-jre (*)
|    |         +--- com.google.code.gson:gson:2.7
|    |         +--- com.google.android:annotations:4.1.1.4
|    |         +--- io.perfmark:perfmark-api:0.19.0
|    |         |    +--- com.google.code.findbugs:jsr305:3.0.2
|    |         |    \--- com.google.errorprone:error_prone_annotations:2.3.3 -> 2.3.4
|    |         +--- io.opencensus:opencensus-api:0.21.0
|    |         \--- io.opencensus:opencensus-contrib-grpc-metrics:0.21.0
|    |              \--- io.opencensus:opencensus-api:0.21.0

//Old dependency tree
+--- project :shared:controller-api
|   +--- org.slf4j:slf4j-api:1.7.25
|   +--- io.grpc:grpc-netty:1.24.2
|    |    +--- io.grpc:grpc-core:1.24.2
|    |    |    +--- io.grpc:grpc-api:1.24.2
|    |    |    |    +--- io.grpc:grpc-context:1.24.2
|    |    |    |    +--- com.google.errorprone:error_prone_annotations:2.3.3 -> 2.3.4
|    |    |    |    +--- com.google.code.findbugs:jsr305:3.0.2
|    |    |    |    +--- org.codehaus.mojo:animal-sniffer-annotations:1.17
|    |    |    |    \--- com.google.guava:guava:26.0-android -> 28.2-jre (*)
|    |    |    +--- com.google.code.gson:gson:2.7
|    |    |    +--- com.google.android:annotations:4.1.1.4
|    |    |    +--- io.perfmark:perfmark-api:0.19.0
|    |    |    |    +--- com.google.code.findbugs:jsr305:3.0.2
|    |    |    |    \--- com.google.errorprone:error_prone_annotations:2.3.3 -> 2.3.4
|    |    |    +--- io.opencensus:opencensus-api:0.21.0
|    |    |    \--- io.opencensus:opencensus-contrib-grpc-metrics:0.21.0
|    |    |         \--- io.opencensus:opencensus-api:0.21.0
|    |    +--- io.netty:netty-codec-http2:4.1.38.Final -> 4.1.50.Final
|    |    |    +--- io.netty:netty-common:4.1.50.Final
|    |    |    +--- io.netty:netty-buffer:4.1.50.Final (*)
|    |    |    +--- io.netty:netty-transport:4.1.50.Final (*)
|    |    |    +--- io.netty:netty-codec:4.1.50.Final (*)
|    |    |    +--- io.netty:netty-handler:4.1.50.Final (*)
|    |    |    \--- io.netty:netty-codec-http:4.1.50.Final
|    |    |         +--- io.netty:netty-common:4.1.50.Final
|    |    |         +--- io.netty:netty-buffer:4.1.50.Final (*)
|    |    |         +--- io.netty:netty-transport:4.1.50.Final (*)
|    |    |         +--- io.netty:netty-codec:4.1.50.Final (*)
|    |    |         \--- io.netty:netty-handler:4.1.50.Final (*)
|    |    \--- io.netty:netty-handler-proxy:4.1.38.Final -> 4.1.50.Final
|    |         +--- io.netty:netty-common:4.1.50.Final
|    |         +--- io.netty:netty-buffer:4.1.50.Final (*)
|    |         +--- io.netty:netty-transport:4.1.50.Final (*)
|    |         +--- io.netty:netty-codec:4.1.50.Final (*)
|    |         +--- io.netty:netty-codec-socks:4.1.50.Final
|    |         |    +--- io.netty:netty-common:4.1.50.Final
|    |         |    +--- io.netty:netty-buffer:4.1.50.Final (*)
|    |         |    +--- io.netty:netty-transport:4.1.50.Final (*)
|    |         |    \--- io.netty:netty-codec:4.1.50.Final (*)
|    |         \--- io.netty:netty-codec-http:4.1.50.Final (*)
```
* Remove explicit dependency of netty-transport from `:shared:protocol` project. 

* grpc-netty-shaded, already includes netty-tcnative on BoringSSL ref:  https://github.com/grpc/grpc-java/blob/master/SECURITY.md

**How to verify it**  
Verify all the tests including system tests pass with and without SSL enabled.

### CODE REVIEW

**Comment by shrids** [comment]
System test run /job/pravega-branch-builder/689/ is in progress...

**Comment by shrids** [comment]
Closing this as there is a separate effort to use on `grpc-okhttp` instead of `grpc-netty`

**Comment by shrids** [comment]
System test is green with this change set test-pravega-continuous-1/1673/

**Review by eolivelli — APPROVED** [review]
LGTM

File: build.gradle
Comment (tkaitchuck) [inline build.gradle:291]: Why is the protocol package being changed also?

File: build.gradle
Comment (tkaitchuck) [inline build.gradle:718]: Here and above, I see the dependency being removed but nothing is added. Where is this shaded dep coming from?

**Review by tkaitchuck — CHANGES_REQUESTED** [review]

File: build.gradle
Comment (shrids) [reply build.gradle:718]: It is coming from the protobuf.gradle . This will be applied fromthe controller-api project.
Let me know if there is a better way.

**Review by derekm — APPROVED** [review]
This PR allowed me to remove these netty overrides in snmp-trap-handler: https://github.com/pravega/snmp-trap-handler/blob/master/snmp-trap-handler/pom.xml#L30

**Review by tkaitchuck — APPROVED** [review]

**Review by eolivelli — APPROVED** [review]

**Comment by shrids** [comment]
The flink longevity test is in progress. Let's wait for the results before this changeset is merged.

**Comment by RaulGracia** [comment]
@shrids are we good to merge this one?

**Comment by shrids** [comment]
The flink longevity results are still not out yet. 
It should be noted that with this PR the properties to configure the netty properties (eg: number of threads, direct memory used) will need to be changed in the downstream projects. It will be prefixed with `io.grpc.netty.shaded.` 
e.g: `io.grpc.netty.shaded.io.netty.eventLoopThreads`  to configure the ELG threads
`io.grpc.netty.shaded.io.netty.maxDirectMemory` to configure the max Direct memory used by netty.

**Comment by RaulGracia** [comment]
@shrids it would be great to create issues and link them to this PR for all the projects impacted by this change before merging. Will block until we have clarity on that.

**Review by RaulGracia — CHANGES_REQUESTED** [review]
Blocking for safety until: i) we identify all the impacted projects as a result of merging this PR to configure Netty settings, and ii) longevity run works successfully.

**Comment by andreipaduroiu** [comment]
@shrids What is the status of this PR? I haven't seen any updates in months. Are we still pursuing it or should we close it?

**Review by crazyzhou — APPROVED** [review]
LGTM. This change can help client users to avoid some netty conflicts.
We have tested this fix against the latest Flink connector, all the connector unit tests can pass.

**Comment by shrids** [comment]
Checked with @crazyzhou the longevity runs were successful with this change set.

**Review by RaulGracia — APPROVED** [review]

**Pull request merged**
